### ISSUE TITLE
Custom lint ?

### ISSUE THREAD

**Issue by giltho** [ocaml/dune#3471]
Hi!

## Desired Behavior

Hello, it seems that the current system makes it impossible to run a custom promotable linting action, as it is possible for ppx.

```dune
(library
   (name ...)
   (lint (pps ppx_yojson_conv))
```

will let you run `dune build @lint` and then `dune promote`

In the same manner, I would expect
```dune
(library
  (name ...)
  (lint (action (echo "test")))
```
to let produce promotable files where every one of my ml and mli file have been overridden by files containing just "test".

I know it's happening in `preprocessing.ml > lint_module > | Action` where the target is set to `None` and no `promote_correction` is happening, but I haven't had time to dive deeper into it to make it work yet

**Comment by jeremiedimino** [comment]
You can, but the protocol is secret :)

If you generate a file called `file.ml.lint-corrected`, dune will automatically offer it as a correction.

**Comment by giltho** [comment]
Unfortunately that does not seem to work :/
That was also my first thought given that this is how it works internally when linting with a `ppx`.

However, it seems that the definition of the promotion is only done in the `Pps` branch of the `lint_module` pattern_matching (see [here](https://github.com/ocaml/dune/blob/9e39a978df39840ee6e4854d07b0a69107c68689/src/dune/preprocessing.ml#L542-L582))

I made a small repro here : https://github.com/giltho/dune-repro-no-lint

I don't know what *should* be the right way of doing that though, writing to .lint-corrected, or doing the same thing as for preprocessing, i.e. using stdout (https://dune.readthedocs.io/en/stable/concepts.html#preprocessing-with-actions).
Given that lint works in the exact same way as preprocessing other wise (only takes an `input-file` parameter, and no target), I would expect the stdout solution

**Comment by jeremiedimino** [comment]
> However, it seems that the definition of the promotion is only done in the Pps branch of the lint_module pattern_matching (see here)

Ah, indeed. I forgot about this.

> I don't know what should be the right way of doing that though, writing to .lint-corrected, or doing the same thing as for preprocessing, i.e. using stdout (https://dune.readthedocs.io/en/stable/concepts.html#preprocessing-with-actions).

I suppose that would work. It's a breaking change though so we'd have to handle it carefully.

@NathanReb has been preparing a new improved "linting pipeline" so that different linters can work well together. So any such change should be considered in the context of this work or at least postponed so that we don't go in different directions.

BTW, if you'd like to help on this front that would be welcome :)

**Comment by giltho** [comment]
Sure I'd love to help with it :) And it's clearly not an emergency so it can wait for the linting pipeline rework to be done

**Comment by jeremiedimino** [comment]
Great :)

If you'd like to help us improve the linting process in general, we can plan a chat between the three of us so that we can show you our plan and discuss how you can help, or if you'd rather focus on the idea of having lint action produce a correction, we'll come back to you once we have the new linting pipeline in place. Just let me know, thanks!

**Comment by giltho** [comment]
Although I'm interested, I already have too much to do on other projects already :/ I think it would be counter productive for me to participate in the linting pipeline improvement. But do ping me once it is done, or almost, so that I can help with that part :)

**Comment by jeremiedimino** [comment]
Sounds good!

**Comment by Kakadu** [comment]
Folks, do you think that specifying linter command should be done outside dune rules, from command line? Imagine the following situation: I force to use dune for homework in my FP course and want to setup a linter. Students remove linting lines from `dune` files and I'm obliged to force right `dune` files administratively, instead simply rinning my linter in CI on all code.
